### PR TITLE
Adds the Node interface to Me and a bunch of optimisations.

### PR DESCRIPTION
**Pull request by sarahscott**

### CODE REVIEW

**Comment by alloy** [comment]
Turns out that for a lot of fields nested under `me` we were making unnecessary gravity `/me` requests in cases where no actual user information was required. In d005a7f25bc9d6421af85b5996dba5446f8e0ae2  we’ve made it so that [these fields](https://github.com/artsy/metaphysics/pull/596/commits/d005a7f25bc9d6421af85b5996dba5446f8e0ae2#diff-f0081da50aef46ac108ae631440ca870R62) alone do not trigger a gravity `/me` request.

/cc @sweir27 @damassi @mzikherman

File: schema/gene.js [hunk 1]
```diff
@@ -99,8 +99,8 @@ const Gene: GraphQLFieldConfig<GeneType, *> = {
   resolve: (root, { id }, request, { fieldNodes }) => {
     // If you are just making an artworks call ( e.g. if paginating )
     // do not make a Gravity call for the gene data.
-    const blacklistedFields = ['filtered_artworks', 'id'];
-    if (!fieldNodes || queriedForFieldsOtherThanBlacklisted(fieldNodes, blacklistedFields)) {
+    const blacklistedFields = ['filtered_artworks', 'id', '__id'];
```
Comment (mzikherman) [inline schema/gene.js]: 👍

File: schema/me/index.js [hunk 2]
```diff
@@ -52,9 +56,23 @@ const Me = new GraphQLObjectType({
...
+      'bidders', 'bidder_positions', 'bidder_status', 'lot_standing', 'lot_standings', 'sale_registrations',
+      'conversations', 'collector_profile',
+      'artwork_inquiries_connection', 'notifications_connection',
+    ];
+    if (queriedForFieldsOtherThanBlacklisted(fieldNodes, blacklistedFields)) {
+      return gravity.with(accessToken)('me')
+        .catch(() => null);
+    }
+
+    // The email and is_collector are here so that the type system's `isTypeOf`
```
Comment (mzikherman) [inline schema/me/index.js]: 👍

**Comment by mzikherman** [comment]
Thanks for this!

I like the commit structure and naming, and the code is very clear.

One thing that the initial `/api/v1/me` API bootstrap call (+ `.catch(() => null)` part especially) did, was swallow any errors and always just return: `{ "data": { "me": null } }` if there was an access token issue, no matter what was requested from under `me` (`lot_standings`, `bidders`, etc.)

Now, the GraphQL response will include the relevant Gravity error message in its `errors` part (such as if the access token is expired or invalid). I think that's good! Rather than silently returning nothing the proper error is. But since its changing the response from Metaphysics in those cases, we should be on the lookout for fallout in clients.

**Comment by mzikherman** [comment]
I'm definitely down to merge this, maybe one of @sweir27 or @damassi can do the honors so they're aware of it as well?

**Comment by alloy** [comment]
> But since this is changing the response from Metaphysics for those cases, we should be on the lookout for fallout in clients.

Good tidbit to be aware of indeed, thanks @mzikherman!

**Comment by sweir27** [comment]
This is a good optimization-- my naive question is whether we can or want to do a larger effort to map requested _fields_ to the endpoint that actually has that data. Maybe instead of explicitly blacklisting fields, we should map each existing field to the endpoint that has that data.

**Comment by alloy** [comment]
@sweir27 Yeah very good question, I’ll have a look see as to how we could extend graphql-js to annotate the fields with that metadata. I won’t make that part of this PR, though.

**Comment by alloy** [comment]
Alright, going for the merge myself so this doesn't get stale and @sarahscott doesn't get blocked.

**Comment by sweir27** [comment]
Sorry! Was out at meetings. Thanks for merging, future PR/discussion is great 👍

**Pull request merged**